TESTS/mbed_drivers/device_key/main.cpp

```diff
@@ -0,0 +1,320 @@
+    NVStore &nvstore = NVStore::get_instance();
+
+    nvstore.init();
+    nvstore.reset();
+    int ret = devkey.device_inject_root_of_trust(key, DEVICE_KEY_32BYTE);
+    TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+
+    //Read the key from NVStore.
+    memset(rkey, 0, DEVICE_KEY_32BYTE);
+    ret = nvstore.get(NVSTORE_KEY_ROT, DEVICE_KEY_32BYTE, rkey, actual_size);
+    TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+    TEST_ASSERT_EQUAL_INT(DEVICE_KEY_32BYTE, actual_size);
+    TEST_ASSERT_EQUAL_INT32_ARRAY(key, rkey, actual_size / sizeof(uint32_t));
+}
+
+/*
+ * Test that a key can be written to persistent memory only once.
+ */
+void device_inject_root_of_trust_only_once_test()
+{
+    DeviceKey &devkey = DeviceKey::get_instance();
+    uint32_t key[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
+    NVStore &nvstore = NVStore::get_instance();
+
+    nvstore.init();
+    int ret = nvstore.reset();
+    TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+
+    memcpy(key, "12345678123456788765432187654321", DEVICE_KEY_32BYTE);
+    ret = devkey.device_inject_root_of_trust(key, DEVICE_KEY_16BYTE);
+    TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+
+    ret = devkey.device_inject_root_of_trust(key, DEVICE_KEY_16BYTE);
+    TEST_ASSERT_EQUAL_INT(DEVICEKEY_ALREADY_EXIST, ret);
+}
+
+/*
+ * Test the consistency of derived 16 byte key result.
+ */
+void device_key_derived_key_consistency_16_byte_key_test()
+{
+    unsigned char output1[DEVICE_KEY_16BYTE];
+    unsigned char output2[DEVICE_KEY_16BYTE];
+    unsigned char empty_buffer[DEVICE_KEY_16BYTE];
+    unsigned char salt[] = { "Once upon a time, I worked for the circus and I lived in Omaha." };
+    int key_type = DEVICE_KEY_16BYTE;
+    DeviceKey &devkey = DeviceKey::get_instance();
+    NVStore &nvstore = NVStore::get_instance();
+
+    nvstore.init();
+    int ret = nvstore.reset();
+    TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+
+    ret = inject_dummy_rot_key();
+    TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+
+    size_t salt_size = sizeof(salt);
+    memset(output1, 0, DEVICE_KEY_16BYTE);
+    ret = devkey.device_key_derived_key(salt, salt_size, output1, key_type);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+    bool is_empty = !memcmp(empty_buffer, output1, sizeof(output1));
+    TEST_ASSERT_FALSE(is_empty);
+
+    for (int i = 0; i < 100; i++) {
+        memset(output2, 0, DEVICE_KEY_16BYTE);
+        ret = devkey.device_key_derived_key(salt, salt_size, output2, key_type);
+        TEST_ASSERT_EQUAL_INT32(0, ret);
+        TEST_ASSERT_EQUAL_UINT8_ARRAY(output1, output2, DEVICE_KEY_16BYTE);
+    }
+}
+
+/*
+ * Test the consistency of derived 32 byte key result.
+ */
+void device_key_derived_key_consistency_32_byte_key_test()
+{
+    unsigned char output1[DEVICE_KEY_32BYTE];
+    unsigned char output2[DEVICE_KEY_32BYTE];
+    unsigned char empty_buffer[DEVICE_KEY_32BYTE];
+    unsigned char salt[] = { "The quick brown fox jumps over the lazy dog" };
```

Comment:
String initialization in this module is not uniform - once with braces, once without. Maybe it's better to lose the braces.